spmd/tensor/parallel/style.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import torch
+from dataclasses import dataclass
+from typing import Callable, Union, Optional
+from spmd.tensor import (
+    DTensor,
+    Shard,
+    Replicate,
+    DeviceMesh,
+)
+
+
+@dataclass
+class ParallelStyle(object):
+    """
+    The parallel style user wants the module or submodule to be sharded.
+    Users can extend this class to build their own parallel style with customized input/output preparations.
+    """
+
+    _prepare_input: Optional[
+        Callable[
+            [Union[torch.Tensor, DTensor], Optional[DeviceMesh], Optional[int]],
+            DTensor,
+        ]
+    ]
+    _prepare_output: Optional[
+        Callable[
+            [DTensor, Optional[DeviceMesh], Optional[int]],
+            Union[torch.Tensor, DTensor],
+        ]
+    ]
+
+
+def shard_input_1d(
+    tensor: Union[torch.Tensor, DTensor],
+    device_mesh: Optional[DeviceMesh] = None,
+    dim: int = 0,
```

Comment:
Now I see the problem. Making dim parameter Optional does not make the param len variable. Now I see the disadvantage of this design. Let's remove Optional[int] and pass it in ParalellStyle. What do you think?